gourmet/timer.py

```diff
@@ -1,100 +1,128 @@
-from gi.repository import Gtk
-from gi.repository import GObject
-import time
-from . import gglobals
 import os
+import time
+from typing import Callable, List, Optional
+
+from gettext import gettext as _
+from gi.repository import Gtk, GObject
 import xml.sax.saxutils
+
+from . import gglobals
 from .sound import Player
 from .gtk_extras import cb_extras as cb
-from gettext import gettext as _
+from .gtk_extras.dialog_extras import getBoolean, UserCancelledError
+
 
 class TimeSpinnerUI:
 
-    def __init__ (self, hoursSpin, minutesSpin, secondsSpin):
-        self.timer_hooks = []
-        self.running = False
-        self.hoursSpin = hoursSpin
-        self.minutesSpin = minutesSpin
-        self.secondsSpin = secondsSpin
-        for s in [self.hoursSpin,self.minutesSpin,self.secondsSpin]:
+    def __init__(self,
+                 hoursSpin: Gtk.SpinButton,
+                 minutesSpin: Gtk.SpinButton,
+                 secondsSpin: Gtk.SpinButton):
+        self.timer_hooks: List[Callable] = []  # actions rand when timer over
+        self.is_running: bool = False  # State flag
+        self.previous_iter_time: Optional[float] = None  # time, used for ticks
+        self.orig_time: int = 0  # in seconds, user input time
+
+        self.hoursSpin: Gtk.SpinButton = hoursSpin
+        self.minutesSpin: Gtk.SpinButton = minutesSpin
+        self.secondsSpin: Gtk.SpinButton = secondsSpin
+
+        for spinner in (self.hoursSpin, self.minutesSpin, self.secondsSpin):
             # This is set up to assure 2 digit entries... 00:00:00, etc.
-            s.connect('changed',self.val_changed_cb)
-            s.val_change_is_changing_entry = False
-            s.set_width_chars(2)
-            s.set_value(1)
-            s.set_value(0)
+            spinner.connect('changed', self.val_changed_cb)
+            spinner.val_change_is_changing_entry = False
+            spinner.set_width_chars(2)
+
+    def set_time(self, val: float) -> None:
+        """Update the spinners on tick.
+
+        Ran on every tick, the value is split into hours, minutes, seconds, and
+        each spinner is set, to show the time going down.
+        """
+        val = max(0., val)
+        self.hoursSpin.set_value(val // 3600)
+        val = val % 3600
+        self.minutesSpin.set_value(val // 60)
+        self.secondsSpin.set_value(round(val % 60))
```

Comment:
Sorry again. I think we might want to truncate the value, rather than round.

Let's say we get a value of `59.8` and `59.8 // 60 == 0.0` (do we need to truncate here?) then that gets rounded to 60 but it probably shouldn't show that. Alternatively, we could just round it to the full second at the beginning of this method (might be the better option).

Reply:
By casting to `int`?

Reply:
yeah, I guess. Let's just do that right at the beginning

Reply:
nice, thanks!